src/lib/operators/table_scan/column_is_null_table_scan_impl.cpp

```diff
@@ -16,46 +24,53 @@ namespace hyrise {
 
 ColumnIsNullTableScanImpl::ColumnIsNullTableScanImpl(const std::shared_ptr<const Table>& in_table,
                                                      const ColumnID column_id,
-                                                     const PredicateCondition& predicate_condition)
-    : _in_table(in_table), _column_id(column_id), _predicate_condition(predicate_condition) {
+                                                     const PredicateCondition& init_predicate_condition)
+    : AbstractDereferencedColumnTableScanImpl(in_table, column_id, init_predicate_condition) {
   DebugAssert(predicate_condition == PredicateCondition::IsNull || predicate_condition == PredicateCondition::IsNotNull,
-              "Invalid PredicateCondition");
+              "Invalid PredicateCondition.");
 }
 
 std::string ColumnIsNullTableScanImpl::description() const {
   return "IsNullScan";
 }
 
-std::shared_ptr<RowIDPosList> ColumnIsNullTableScanImpl::scan_chunk(const ChunkID chunk_id) {
-  const auto& chunk = _in_table->get_chunk(chunk_id);
-  const auto& segment = chunk->get_segment(_column_id);
-
-  auto matches = std::make_shared<RowIDPosList>();
-
-  if (const auto value_segment = std::dynamic_pointer_cast<BaseValueSegment>(segment)) {
-    _scan_value_segment(*value_segment, chunk_id, *matches);
-  } else {
-    const auto& chunk_sorted_by = chunk->individually_sorted_by();
-    if (!chunk_sorted_by.empty()) {
-      for (const auto& sorted_by : chunk_sorted_by) {
-        if (sorted_by.column == _column_id) {
-          _scan_generic_sorted_segment(*segment, chunk_id, *matches, sorted_by.sort_mode);
-          ++num_chunks_with_binary_search;
-          return matches;
+void ColumnIsNullTableScanImpl::_scan_non_reference_segment(
+    const AbstractSegment& segment, const ChunkID chunk_id, RowIDPosList& matches,
+    const std::shared_ptr<const AbstractPosList>& position_filter) {
+  resolve_data_type(segment.data_type(), [&](auto type) {
+    using SegmentDataType = typename decltype(type)::type;
+
+    // The ColumnIsNullTableScan is optimized for Value, Dictionary, LZ4, and FrameofReference segments since their
+    // NULL values can be efficiently iterated through their null_values or attribute vector. RunLength segments use
+    // the _scan_generic_segment() method because their NULL values are stored in runs, making iteration less
+    // straightforward.
+
+    if (!(_try_non_generic_segment_scan<BaseValueSegment>(segment, chunk_id, matches, position_filter) ||
+          _try_non_generic_segment_scan<BaseDictionarySegment>(segment, chunk_id, matches, position_filter) ||
+          _try_non_generic_segment_scan<LZ4Segment<SegmentDataType>>(segment, chunk_id, matches, position_filter) ||
+          _try_non_generic_segment_scan<FrameOfReferenceSegment<int32_t>>(segment, chunk_id, matches,
```

Comment:
Does anybody know how to do without using `int32_t`?

Reply:
Nope. At least nothing that is worth the effort.